src/ThreadPool.cc

```diff
@@ -0,0 +1,327 @@
+/**
+ * \file    ThreadPool.cc
+ * \authors \ref contributors
+ * \date    Created on November 6, 2018
+ * \brief   Implementation of the ThreadPool class.
+ *
+ * \copyright Copyright 2018-2024 Deutsches Elektronen-Synchrotron (DESY), Hamburg
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation, either version 2.1 of the license, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+#include <algorithm>
+#include <iostream>
+
+#include <gul14/cat.h>
+#include <gul14/ThreadPool.h>
+#include <gul14/time_util.h>
+
+#if defined(__APPLE__) || defined(__GNUC__)
+#include <signal.h>
+#endif
+
+namespace gul14 {
+
+ThreadPool::ThreadPool(std::size_t num_threads, std::size_t capacity)
+    : capacity_(capacity)
+{
+    if (num_threads < 1 || num_threads > max_threads)
+    {
+        throw std::invalid_argument(
+            cat("Illegal number of threads for thread pool: ", num_threads));
+    }
+
+    if (capacity < 1 || capacity > max_capacity)
+        throw std::invalid_argument(cat("Illegal capacity for thread pool: ", capacity));
+
+    threads_.reserve(num_threads);
+    while (threads_.size() < num_threads)
+        threads_.emplace_back([this]() { perform_work(); });
```

Comment:
:thinking: why is this not directly?

```suggestion
        threads_.emplace_back(&perform_work);
```

Maybe lack of :coffee: on my side.

Reply:
`perform_work()` is a member function, so we need to capture the `this` pointer for it. We could also make it a free function, but then we need to pass a reference to the `ThreadPool` to it. I figure this is simpler.